File: packages/sp_shared/lib/src/resolvers/year_resolver.dart
import '../models/episode_data.dart';
import '../models/smart_playlist.dart';
import '../models/smart_playlist_definition.dart';
import '../models/smart_playlist_sort.dart';
import '../models/smart_playlist_title_extractor.dart';
import 'smart_playlist_resolver.dart';

/// Resolver that groups episodes by publication year.
class YearResolver implements SmartPlaylistResolver {
  @override
  String get type => 'year';

  @override
  SmartPlaylistSortSpec get defaultSort => const SmartPlaylistSortSpec([
    SmartPlaylistSortRule(
      field: SmartPlaylistSortField.playlistNumber,
      order: SortOrder.descending, // Newest years first
    ),
  ]);

  @override
  SmartPlaylistGrouping? resolve(
    List<EpisodeData> episodes,
    SmartPlaylistDefinition? definition,
  ) {
    final grouped = <int, List<EpisodeData>>{};
    final ungrouped = <int>[];

    for (final episode in episodes) {
      final pubDate = episode.publishedAt;
      if (pubDate != null) {
        grouped.putIfAbsent(pubDate.year, () => []).add(episode);
      } else {
        ungrouped.add(episode.id);
      }
    }

    // Return null if no episodes have publish dates
    if (grouped.isEmpty) {
      return null;
    }

    final titleExtractor = definition?.titleExtractor;

    final playlists = grouped.entries.map((entry) {
      final playlistEpisodes = entry.value;
      final displayName = _extractDisplayName(
        year: entry.key,
        episodes: playlistEpisodes,
        titleExtractor: titleExtractor,
      );

      return SmartPlaylist(
        id: 'year_${entry.key}',
        displayName: displayName,
        sortKey: entry.key,
        episodeIds: playlistEpisodes.map((e) => e.id).toList(),
      );
    }).toList()..sort((a, b) => b.sortKey.compareTo(a.sortKey)); // Descending

    return SmartPlaylistGrouping(
      playlists: playlists,
      ungroupedEpisodeIds: ungrouped,
      resolverType: type,
    );
  }

  String _extractDisplayName({
    required int year,
    required List<EpisodeData> episodes,
    required SmartPlaylistTitleExtractor? titleExtractor,
  }) {
    if (titleExtractor == null || episodes.isEmpty) {
      return '$year';
    }

    // Try to extract title from first episode
    final extracted = titleExtractor.extract(episodes.first);
    return extracted ?? '$year';
  }
}

File: crates/sp_core/src/resolvers/year_resolver.rs
use std::collections::BTreeMap;

use chrono::Datelike;

use crate::models::{
    EpisodeData, Grouping, Playlist, PlaylistDefinition, SortField, SortOrder, SortRule,
};

use super::resolver::Resolver;

/// Resolver that groups episodes by publication year.
pub struct YearResolver;

impl Resolver for YearResolver {
    fn resolver_type(&self) -> &str {
        "year"
    }

    fn default_sort(&self) -> SortRule {
        SortRule {
            field: SortField::PlaylistNumber,
            order: SortOrder::Descending, // Newest years first
        }
    }

    fn resolve(
        &self,
        episodes: &[&dyn EpisodeData],
        definition: Option<&PlaylistDefinition>,
    ) -> Option<Grouping> {
        let title_extractor = definition.and_then(|d| d.title_extractor.as_ref());

        let mut grouped: BTreeMap<i32, Vec<&dyn EpisodeData>> = BTreeMap::new();
        let mut ungrouped: Vec<i64> = Vec::new();

        for &episode in episodes {
            match episode.published_at() {
                Some(pub_date) => {
                    let year = pub_date.year();
                    grouped.entry(year).or_default().push(episode);
                }
                None => {
                    ungrouped.push(episode.id());
                }
            }
        }

        if grouped.is_empty() {
            return None;
        }

        let mut playlists: Vec<Playlist> = grouped
            .iter()
            .map(|(&year, eps)| {
                let display_name =
                    super::extract_display_name_with_fallback(
                        year.to_string(),
                        eps,
                        title_extractor,
                    );
                Playlist::new(
                    format!("year_{}", year),
                    display_name,
                    year,
                    eps.iter().map(|e| e.id()).collect(),
                )
            })
            .collect();

        // Sort by year descending (newest first)
        playlists.sort_by(|a, b| b.sort_key.cmp(&a.sort_key));

        Some(Grouping {
            playlists,
            ungrouped_episode_ids: ungrouped,
            resolver_type: self.resolver_type().to_string(),
        })
    }
}

